File: backend/shared/combat_events.py
from __future__ import annotations

from collections.abc import Iterable, Mapping
from typing import Any


def build_team_lookup(players: Iterable[dict[str, Any]] | None) -> dict[str, str]:
    teams: dict[str, str] = {}
    for player in players or []:
        puuid = str(player.get("puuid") or "").strip()
        team_id = str(player.get("teamId") or "").strip()
        if puuid and team_id:
            teams[puuid] = team_id
    return teams


def is_valid_kill(
    kill: Mapping[str, Any] | None,
    team_by_puuid: Mapping[str, str] | None = None,
) -> bool:
    if not kill:
        return False
    killer = str(kill.get("killer") or "").strip()
    victim = str(kill.get("victim") or "").strip()
    if not killer or not victim or killer == victim:
        return False

    teams = team_by_puuid or {}
    killer_team = teams.get(killer)
    victim_team = teams.get(victim)
    return not (killer_team and victim_team and killer_team == victim_team)


def valid_kills(
    kills: Iterable[dict[str, Any]] | None,
    team_by_puuid: Mapping[str, str] | None = None,
) -> list[dict[str, Any]]:
    return [kill for kill in kills or [] if is_valid_kill(kill, team_by_puuid)]


def is_enemy_damage(
    attacker_puuid: Any,
    receiver_puuid: Any,
    team_by_puuid: Mapping[str, str] | None = None,
) -> bool:
    attacker = str(attacker_puuid or "").strip()
    receiver = str(receiver_puuid or "").strip()
    if not attacker or not receiver or attacker == receiver:
        return False

    teams = team_by_puuid or {}
    attacker_team = teams.get(attacker)
    receiver_team = teams.get(receiver)
    return not (
        attacker_team
        and receiver_team
        and attacker_team == receiver_team
    )


def valid_assistants(
    kill: Mapping[str, Any] | None,
    team_by_puuid: Mapping[str, str] | None = None,
) -> list[str]:
    if not is_valid_kill(kill, team_by_puuid):
        return []

    killer = str(kill.get("killer") or "").strip()
    victim = str(kill.get("victim") or "").strip()
    teams = team_by_puuid or {}
    killer_team = teams.get(killer)
    result: list[str] = []
    seen: set[str] = set()
    for value in kill.get("assistants") or []:
        assistant = str(value or "").strip()
        if (
            not assistant
            or assistant in seen
            or assistant == killer
            or assistant == victim
        ):
            continue
        assistant_team = teams.get(assistant)
        if killer_team and assistant_team and assistant_team != killer_team:
            continue
        seen.add(assistant)
        result.append(assistant)
    return result

File: frontend/src/utils/stats/combatEvents.ts
import type {
  RawKillEvent,
  RawPlayer,
  RawRound,
  RawRoundDamage,
} from "../../types/matches";

export type TeamLookup = ReadonlyMap<string, string>;

const id = (value: unknown) => String(value ?? "").trim();

export function buildTeamLookup(players: RawPlayer[] | undefined): Map<string, string> {
  const teams = new Map<string, string>();
  for (const player of players ?? []) {
    const puuid = id(player.puuid);
    const teamId = id(player.teamId);
    if (puuid && teamId) teams.set(puuid, teamId);
  }
  return teams;
}

export function isValidKill(
  kill: RawKillEvent | null | undefined,
  teams?: TeamLookup,
  fallbackKiller?: string,
): boolean {
  const killer = id(kill?.killer) || id(fallbackKiller);
  const victim = id(kill?.victim);
  if (!killer || !victim || killer === victim) return false;
  const killerTeam = teams?.get(killer);
  const victimTeam = teams?.get(victim);
  return !(killerTeam && victimTeam && killerTeam === victimTeam);
}

export function isEnemyDamage(
  attacker: unknown,
  damage: RawRoundDamage,
  teams?: TeamLookup,
): boolean {
  const attackerId = id(attacker);
  const receiverId = id(damage.receiver);
  if (!attackerId || !receiverId || attackerId === receiverId) return false;
  const attackerTeam = teams?.get(attackerId);
  const receiverTeam = teams?.get(receiverId);
  return !(attackerTeam && receiverTeam && attackerTeam === receiverTeam);
}

export function validAssistants(
  kill: RawKillEvent,
  teams?: TeamLookup,
  fallbackKiller?: string,
): string[] {
  if (!isValidKill(kill, teams, fallbackKiller)) return [];
  const killer = id(kill.killer) || id(fallbackKiller);
  const victim = id(kill.victim);
  const killerTeam = teams?.get(killer);
  const raw = Array.isArray(kill.assistants)
    ? kill.assistants
    : typeof kill.assistants === "string"
      ? [kill.assistants]
      : [];
  return [...new Set(raw.map(id))].filter((assistant) => {
    if (!assistant || assistant === killer || assistant === victim) return false;
    const assistantTeam = teams?.get(assistant);
    return !(killerTeam && assistantTeam && killerTeam !== assistantTeam);
  });
}

export type CollectedRoundKill = {
  kill: RawKillEvent;
  ownerPuuid: string;
};

export function collectRoundKills(round: RawRound): CollectedRoundKill[] {
  const seen = new Set<string>();
  const kills: CollectedRoundKill[] = [];
  for (const stat of round.playerStats ?? []) {
    const ownerPuuid = id(stat.puuid);
    for (const kill of stat.kills ?? []) {
      const key = [
        kill.timeSinceRoundStartMillis ?? "",
        id(kill.killer) || ownerPuuid,
        id(kill.victim),
        validRawAssistants(kill).sort().join(","),
      ].join("|");
      if (seen.has(key)) continue;
      seen.add(key);
      kills.push({ kill, ownerPuuid });
    }
  }
  return kills.sort(
    (a, b) =>
      Number(a.kill.timeSinceRoundStartMillis ?? 0) -
      Number(b.kill.timeSinceRoundStartMillis ?? 0),
  );
}

function validRawAssistants(kill: RawKillEvent): string[] {
  if (Array.isArray(kill.assistants)) return kill.assistants.map(id).filter(Boolean);
  return typeof kill.assistants === "string" && id(kill.assistants)
    ? [id(kill.assistants)]
    : [];
}
